### ISSUE TITLE
Executable detection in Linux

### ISSUE THREAD

**Issue by nauxiew** [almarklein/pyelastix#18]
After reading the web and github of elastix, I still can not figure out how to install the Elastix command line application. And an error occurred to me, i.e. **[Errno 2] No such file or directory: '/opt/local/bin'**
I tried to install SimpleElastix, but lots of error occurred, so I gave up and tried to use pyelastix.

**Comment by almarklein** [comment]
I think its a matter of downloading a zip-file and putting it in a sensible place :)

**Comment by nauxiew** [comment]
> I think its a matter of downloading a zip-file and putting it in a sensible place :)

Sorry, I don't know what "a sensible place" mean? could you give me an example, really thanks for your reply

**Comment by almarklein** [comment]
Good point :)  If you're on Windows, that'd be `c:\program files` (https://github.com/almarklein/pyelastix/blob/master/pyelastix.py#L98)

**Comment by nauxiew** [comment]
> Good point :) If you're on Windows, that'd be `c:\program files` (https://github.com/almarklein/pyelastix/blob/master/pyelastix.py#L98)

thanks, what if I am in linux system, does it mean that I should do some modifications like: ### sys.platform.startswith('linux'), and indicate a specific path like: ### "/my_account/bin/", and download elastix in this path. I will try it, really thanks for your reply

**Comment by nauxiew** [comment]
> Good point :) If you're on Windows, that'd be `c:\program files` (https://github.com/almarklein/pyelastix/blob/master/pyelastix.py#L98)

Sorry, I occurred a new error, I solve the former problem by doing these:
1. I downloaded the elastix from http://elastix.isi.uu.nl/download_links.php, I clicked " **Download**" and  downloaded binaries via GitHub (linux version), and unzip it to **"/my_account/elastix"**
2. I modified  **pyelastix.py#L102** as possible_locations.extend([ "/my_account/elastix/bin"]), since elastix and transform files are in ./bin/
3. a new error occurred: **Could not find Elastix executable. Download Elastix from http://elastix.isi.uu.nl/. Pyelastix looks for the exe in a series of common locations. Set ELASTIX_PATH if necessary.**

To solve it, I try this code: **os.environ['ELASTIX_PATH'] = r'/my_account/elastix/bin/'** before excuting pyelastix.register(), however, unsuccessful.
Then, I downloaded win version from GitHub, I found there were two files: elastix.exe and transform.exe, but the unzipped linux version files are: elastix and transform, without .exe, (I download elastix from https://github.com/SuperElastix/elastix/releases/tag/4.9.0), I don't know how to solve it.

**Comment by zwep** [comment]
> 
> 
> > Good point :) If you're on Windows, that'd be `c:\program files` (https://github.com/almarklein/pyelastix/blob/master/pyelastix.py#L98)
> 
> Sorry, I occurred a new error, I solve the former problem by doing these:
> 
>     1. I downloaded the elastix from http://elastix.isi.uu.nl/download_links.php, I clicked " **Download**" and  downloaded binaries via GitHub (linux version), and unzip it to **"/my_account/elastix"**
> 
>     2. I modified  **pyelastix.py#L102** as possible_locations.extend([ "/my_account/elastix/bin"]), since elastix and transform files are in ./bin/
> 
>     3. a new error occurred: **Could not find Elastix executable. Download Elastix from http://elastix.isi.uu.nl/. Pyelastix looks for the exe in a series of common locations. Set ELASTIX_PATH if necessary.**
> 
> 
> To solve it, I try this code: **os.environ['ELASTIX_PATH'] = r'/my_account/elastix/bin/'** before excuting pyelastix.register(), however, unsuccessful.
> Then, I downloaded win version from GitHub, I found there were two files: elastix.exe and transform.exe, but the unzipped linux version files are: elastix and transform, without .exe, (I download elastix from https://github.com/SuperElastix/elastix/releases/tag/4.9.0), I don't know how to solve it.

To remove your last confusion.. windows uses .exe-files to denote executables, unix system won't use exentions like that. So that seems fine I guess

**Comment by nauxiew** [comment]
> > > Good point :) If you're on Windows, that'd be `c:\program files` (https://github.com/almarklein/pyelastix/blob/master/pyelastix.py#L98)
> > 
> > 
> > Sorry, I occurred a new error, I solve the former problem by doing these:
> > ```
> > 1. I downloaded the elastix from http://elastix.isi.uu.nl/download_links.php, I clicked " **Download**" and  downloaded binaries via GitHub (linux version), and unzip it to **"/my_account/elastix"**
> > 
> > 2. I modified  **pyelastix.py#L102** as possible_locations.extend([ "/my_account/elastix/bin"]), since elastix and transform files are in ./bin/
> > 
> > 3. a new error occurred: **Could not find Elastix executable. Download Elastix from http://elastix.isi.uu.nl/. Pyelastix looks for the exe in a series of common locations. Set ELASTIX_PATH if necessary.**
> > ```
> > 
> > 
> > To solve it, I try this code: **os.environ['ELASTIX_PATH'] = r'/my_account/elastix/bin/'** before excuting pyelastix.register(), however, unsuccessful.
> > Then, I downloaded win version from GitHub, I found there were two files: elastix.exe and transform.exe, but the unzipped linux version files are: elastix and transform, without .exe, (I download elastix from https://github.com/SuperElastix/elastix/releases/tag/4.9.0), I don't know how to solve it.
> 
> To remove your last confusion.. windows uses .exe-files to denote executables, unix system won't use exentions like that. So that seems fine I guess

Thanks, after I unzip elastix linux version, **transform** and **elastix** (2 files) are in ./bin, and I indicate the path but program error told me there is no executable Elastix, I am really confused.

**Comment by almarklein** [comment]
My advice would be to add a print statement [here](https://github.com/almarklein/pyelastix/blob/master/pyelastix.py#L105) (and decomment the one a few lines down), so you can see what files are attempted.

**Comment by soupault** [comment]
On Linux, one needs to set `ELASTIX_PATH` to full path to `elastix` binary (`.../elastix-4.9.0-linux/bin/elastix`).

Also, the distributed libraries must be available from the common locations (e.g. `/usr/lib`). Alternatively, set/add to `LD_LIBRARY_PATH` the path to `lib` directory of `elastix` (`.../elastix-4.9.0-linux/lib/`).

---

The issue with `[Errno 2] No such file or directory: '/opt/local/bin'` is a duplicate of #15 and will be closed with #20 .

**Comment by JepsonNomad** [comment]
How should this be set for MacOS? I have installed elastix using method 3.2.1 ("The really easy way") in the [manual](https://elastix.lumc.nl/download/elastix-5.0.0-manual.pdf) and added the PATH and DYLD_LIBRARY_PATH info to my .bash_profile file. I can run `$ elastix --help` in Terminal just fine. However, with Python, `pyelastix.register()` fails with the error 

> RuntimeError: Could not find Elastix executable. Download Elastix from http://elastix.isi.uu.nl/. Pyelastix looks for the exe in a series of common locations. Set ELASTIX_PATH if necessary.

I tried explicitly setting the ELASTIX_PATH environment parameter using 
`os.environ["ELASTIX_PATH"] = "/Applications/elastix/bin/elastix"
` 
(which is where elastix lives according to `$ which elastix` in Terminal).

Mac OS Catalina 10.15.3
Python 3.7.4
elastix 5.0.0
pyelastix 1.2

**Comment by horsto** [comment]
@JepsonNomad same problem here. MacOS Catalina 10.15.6. 
Did you find any solution for this?
```
RuntimeError: Could not find Elastix executable. 
Download Elastix from http://elastix.isi.uu.nl/. Pyelastix looks for the exe in a series of common locations.
Set ELASTIX_PATH if necessary.
```
`which elastix` works fine, and setting the `os.environ['ELASTIX_PATH']` manually also still runs into the same error message.

**Comment by JepsonNomad** [comment]
@horsto Sadly I gave up on pyelastix after juggling some other projects. Good luck, friend.

**Comment by horsto** [comment]
Thanks @JepsonNomad. 

Any idea @soupault ?

**Comment by basicskywards** [comment]
> On Linux, one needs to set `ELASTIX_PATH` to full path to `elastix` binary (`.../elastix-4.9.0-linux/bin/elastix`).
> 
> Also, the distributed libraries must be available from the common locations (e.g. `/usr/lib`). Alternatively, set/add to `LD_LIBRARY_PATH` the path to `lib` directory of `elastix` (`.../elastix-4.9.0-linux/lib/`).
> 
> The issue with `[Errno 2] No such file or directory: '/opt/local/bin'` is a duplicate of #15 and will be closed with #20 .

I also encounter this issue:
**Could not find Elastix executable. Download Elastix from http://elastix.isi.uu.nl/. Pyelastix looks for the exe in a series of common locations. Set ELASTIX_PATH if necessary.**

For Linux, the problem is solved when i set paths as below. Hope it useful for someone.
```
os.environ['LD_LIBRARY_PATH'] = '/path/lib'
os.environ['ELASTIX_PATH']= "/path/bin/elastix"
```

**Comment by gaolanmei** [comment]
# 在terminal：
# cd /home/anaconda3/envs/lib/python3.11/site-packages
# wget https://github.com/SuperElastix/elastix/releases/download/5.0.0/elastix-5.0.0-linux.tar.bz2 #下载失败；从官网下载上传至site-packages
# 注意此处的版本elastix-5.0.0-linux，因为系统版本gcc version 7.5.0 (Ubuntu 7.5.0-3ubuntu1~18.04)，elastix-5.0.0-linux的gcc 7.4.0
# tar -xvjf elastix-5.0.1-linux.tar.bz2
# export PATH=$PATH:/home/anaconda3/envs/lib/python3.11/site-packages/elastix-5.0.0-linux/bin
# which elastix

# 在shell中：
# echo 'export PATH=$PATH:/home/anaconda3/envs/lib/python3.11/site-packages/elastix-5.0.0-linux/bin' >> ~/.bashrc
# source ~/.bashrc
# echo 'export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:/home/anaconda3/envs/lib/python3.11/site-packages/elastix-5.0.0-linux/lib' >> ~/.bashrc
# source ~/.bashrc

os.environ['LD_LIBRARY_PATH'] = '/home/anaconda3/envs/lib/python3.11/site-packages/elastix-5.0.0-linux/lib'
os.environ['ELASTIX_PATH']= "/home/anaconda3/envs/lib/python3.11/site-packages/elastix-5.0.0-linux/bin/elastix"

# 配准参数设置
params = pyelastix.get_default_params(type='RIGID') #type='BSPLINE','RIGID', 'AFFINE', 'BSPLINE'
params.NumberOfResolutions = 4  # 多分辨率策略
params.MaximumNumberOfIterations = 500  # 每级最大迭代次数

print("mov→ref:", mov,ref)

# 执行配准
result,fields = pyelastix.register(mov,ref, params)

if isinstance(result, pyelastix.Image):
    result_array = np.array(result)
    result = sitk.GetImageFromArray(result_array)
    result.CopyInformation(ref_image)
# 保存配准结果
output_filename = mov.replace('.nii.gz', '_r.nii.gz')
print("Saving result to:", output_filename)
sitk.WriteImage(result, output_filename)
